File: elixir/day3/day3.exs
defmodule Aoc2025.Day3 do

  def part1(input) do
    input
      |> Enum.map(fn bank ->
        digits = Enum.map(bank, fn digit -> 
          String.to_integer(digit)
        end)
          |> Enum.with_index()

        max_1 = digits
          |> Enum.take(length(digits) - 1)
          |> Enum.max_by(& elem(&1, 0))
        max_2 = digits
          |> Enum.drop(elem(max_1, 1) + 1)
          |> Enum.max_by(& elem(&1, 0))

        (10 * elem(max_1, 0)) + elem(max_2, 0)
      end)
      |> Enum.sum
  end

  def part2(input) do
    input
      |> Enum.map(fn bank ->
        digits = Enum.map(bank, fn digit -> 
          String.to_integer(digit)
        end)
          |> Enum.with_index()

        {_, _, _, number} = 
          Enum.reduce(digits, {0, length(digits)-11, 11, 0}, fn {digit, idx}, {from, to, exponent, number} ->
            if idx >= from do
              max = Enum.max_by(digits |> Enum.drop(from) |> Enum.take(to-from), & elem(&1, 0))
              {elem(max, 1) + 1, to + 1, exponent - 1, number + trunc(:math.pow(10, exponent)) * elem(max, 0) }
            else
              {from, to, exponent, number}
            end
          end)

        number
      end)
      |> Enum.sum
  end
  
end

# input = "987654321111111
# 811111111111119
# 234234234234278
# 818181911112111"
input = File.read!("input.txt")
  |> String.split("\n")
  |> Enum.filter(& String.length(&1) > 0)
  |> Enum.map(& String.graphemes/1)

sol1 = Aoc2025.Day3.part1(input)
sol2 = Aoc2025.Day3.part2(input)

IO.puts "Part 1: #{sol1}"
IO.puts "Part 2: #{sol2}"

File: cpp/day3/day3.cpp
#include <algorithm>
#include <numeric>
#include <span> 
#include <string>
#include <fstream>
#include <vector>
#include <ranges>
#include <print>
#include <math.h>

int part1(std::span<std::vector<int>> input) {
   return std::transform_reduce(input.cbegin(), input.cend(), 0, std::plus{}, [](std::span<int const> bank) {
      auto left = bank.cbegin();
      auto right = std::prev(bank.cend(), 1);

      auto const max_1 = std::ranges::max_element(left, right);
      auto const max_2 = std::ranges::max_element(std::next(max_1), bank.cend());

      return (10 * *max_1) + *max_2;
   });
}

long part2(std::span<std::vector<int>> input) {
   return std::transform_reduce(input.cbegin(), input.cend(), 0L, std::plus{}, [](std::span<int const> bank) {
      auto left = bank.cbegin();
      auto right = std::prev(bank.cend(), 11);
      auto exponent = 11;
      long number{};

      while (exponent >= 0) {
         auto const max = std::ranges::max_element(left, right);
         number += std::pow(10, exponent--) * *max;
         left = std::next(max);
         right = std::next(right);
      }

      return number;
   });
}

int main() {

   using namespace std::literals;

//    auto input = "987654321111111\n\
// 811111111111119\n\
// 234234234234278\n\
// 818181911112111"sv;


   std::ifstream f{ "input.txt" };
   std::string input{ std::istreambuf_iterator<char>{ f }, std::istreambuf_iterator<char>{} };
   f.close();

   auto v = std::ranges::views::split(input, '\n')
          | std::ranges::views::transform([](auto&& rng) {
               return std::string_view(rng.begin(), rng.end());
            })
          | std::ranges::views::filter([](auto sv) {
            return not sv.empty();
          })
          | std::ranges::views::transform([](std::string_view str) {
               auto ret = std::vector<int>{};
               ret.reserve(str.length());
               for (auto ch : str) {
                  ret.push_back(ch - 48);
               }
               return ret;
            })
            | std::ranges::to<std::vector>();

   auto const sol1 = part1(v);
   auto const sol2 = part2(v);
               
   std::print("Part 1: {0}\n", sol1);
   std::print("Part 2: {0}\n", sol2);

   return 0;
}
